[SPARK-11195][CORE] Use correct classloader for TaskResultGetter

choochootrain opened the pull request:
Make sure we are using the context classloader when deserializing failed TaskResults instead of the Spark classloader.

Commit:
[SPARK-11195][CORE] Use correct classloader for TaskResultGetter

Make sure we are using the context classloader when deserializing failed
TaskResults instead of the Spark classloader.

JoshRosen commented:
Jenkins, this is ok to test.

AmplabJenkins commented:
Can one of the admins verify this patch?

choochootrain commented:
I have a manual test that exhibits this behavior in https://issues.apache.org/jira/browse/SPARK-11195 and I am working on adding a test to the repo.

In order to test this I basically want to mirror the classloader hierarchy created by `spark-submit` - are there any conventions or existing tests which do something like this that I can look at?

JoshRosen commented:
Jenkins, this is ok to test.

JoshRosen commented:
@brkyvz might know more about testing Spark Submit.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #44655 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44655/consoleFull) for   PR 9367 at commit [`90c47aa`](https://github.com/apache/spark/commit/90c47aa133de83dd49dea00b0949088e8600ab23).

yhuai commented:
Will `SparkSubmitSuite` help?

choochootrain commented:
`SparkSubmitSuite` is helpful, but I want to catch and assert the type of exception that is thrown when the job fails - calling into `doMain` in `SparkSubmit.scala` would be closer?

yhuai commented:
How about this. In the main method, you can catch the exception and if it is the expected type, let the main method finish successfully. Otherwise, throw an exception, which causes a non-zero exit code. In `runSparkSubmit`, we will find that the exit code is not 0 and fail the test.

SparkQA commented:
[Test build #44655 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44655/console) for   PR 9367 at commit [`90c47aa`](https://github.com/apache/spark/commit/90c47aa133de83dd49dea00b0949088e8600ab23).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44655/
Test PASSed.

srowen commented:
This looks good as I'm not otherwise clear why these two blocks would use different classloaders.

srowen commented:
@choochootrain are you able to put together a small test like what @yhuai mentions? then I think this is good to go.

choochootrain commented:
i'm writing the test right now, is there an easy way to get the relative path to the assembled spark jar so I can compile my job against it?

yhuai commented:
@choochootrain I think you can write your app in SparkSubmitSuite (https://github.com/apache/spark/blob/master/core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala#L590-L614 is an example).

choochootrain commented:
@yhuai i can't write the test directly in `SparkSubmitSuite` because this is a classloader issue that only repros (as far as I can tell) when an external jar is loaded. I'm adding a test that uses `TestUtils.createCompiledClass` to compile and submit my external jar.

yhuai commented:
@choochootrain I just noticed that this PR is for branch-1.5. Should we also fix it in master?

choochootrain commented:
@yhuai yep this should also be in master. should I also submit a pr on master when this one looks good or will a maintainer be able to cherry-pick the commit?

yhuai commented:
@choochootrain It will be great if you can submit a pr against the master. Our merge script only cherry-pick commit from master to a branch.

Commit:
[SPARK-11195][CORE] Add SparkSubmitSuite test for classloader

This test compiles an external Spark job which throws a user defined
exception and asserts that Spark handles the TaskResult deserialization
properly.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #44987 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44987/consoleFull) for   PR 9367 at commit [`c63ca09`](https://github.com/apache/spark/commit/c63ca091714c464f22e9e28b8605aa8c551eaa87).

SparkQA commented:
[Test build #44987 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44987/console) for   PR 9367 at commit [`c63ca09`](https://github.com/apache/spark/commit/c63ca091714c464f22e9e28b8605aa8c551eaa87).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44987/
Test FAILed.

choochootrain commented:
whoops, fixed the style error.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

choochootrain commented on core/src/test/scala/org/apache/spark/deploy/SparkSubmitSuite.scala:
```diff
@@ -366,6 +367,72 @@ class SparkSubmitSuite
+        |}
+      """.stripMargin)
+    // scalastyle:on line.size.limit
+    val sparkJar = "../assembly/target/scala-2.10/spark-assembly-1.5.1-hadoop2.2.0.jar"
```
the only remaining issue is that this jar is hardcoded for the maven profiles that i am using. i was experimenting with putting the entire maven `target/classes` directory on the classpath but that seems equally janky. any suggestions here?

yhuai replied:
I probably missed something. Why do we need to put spark's assembly jar at here? When we run test, spark's classes are already in class path.

yhuai replied:
btw, when you unit test `SparkSubmitSuite`, you have to do `assembly/assembly` before you can run any of these tests.

yhuai replied:
This is what our jenkins does.

choochootrain replied:
in order to test this issue, i need to submit an external jar to `spark-submit`. i can compile my job it using `TestUtils`, but i need to put (some version) of spark on the classpath so that `javac` can resolve `RDD` and so on.

choochootrain replied:
so the test `classes are correctly loaded when tasks fail` is failing because of the above issue: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44991/consoleFull

i'm more than happy to take another approach if i'm going about this the wrong way.

yhuai replied:
@JoshRosen @andrewor14 any idea on how to test this?

brkyvz replied:
Why do you need to compile using Spark? Can't you just create your own exception, compile it, pass that to Spark Submit. Then move all of the code here down to something like `SimpleApplicationTest`, where you create your exception through reflection and then throw it?

brkyvz replied:
Or maybe I didn't understand the issue that this patch is trying to solve very well

yhuai replied:
Maybe you can compile a jar that just include your jar and put your app below. In your app, you use reflection to create an instance of your exception.

choochootrain replied:
all this patch does is make `TaskResult` deserialization use `Utils.getContextOrSparkClassLoader` (the classloader which loaded the `spark-submit`ed jar) instead of `Utils.getSparkClassLoader` (this is `AppClassLoader` which only has spark classes in it). without this patch, a failed task would not be able to deserialize an exception if it did not exist in `Utils.getSparkClassLoader`.

in order to reproduce this issue, i set up a situation where `Utils.getContextOnSparkClassLoader` contains `MyException` but `Utils.getSparkClassLoader` does not (see https://issues.apache.org/jira/browse/SPARK-11195). this is easy to manually test with `spark-submit` and a user defined exception, but turning this into an automated test is proving to be much trickier. here are the 3 options:

* :x: if i place all of the code into `SparkSubmitSuite`, the bug won't be hit because `MyException` will be in the root classloader and my patch makes no difference.

* :grey_question: if i place all of the code into an external jar and run `spark-submit`, i can set up the same situation as my repro which found this bug. the issue i am running into is that i need a spark classpath in order to compile my jar. i can use the assembled jar, but this changes depending on the maven profiles that are enabled and so on.

* :grey_question: i can try @brkyvz & @yhuai's hybrid approach of putting only the exception into a jar and the rest of the code into `SparkSubmitSuite`. i will have to do the following in order to repro this issue:
  * load the jar with `MyException` in a new classloader whose parent is the root classloader
  * somehow allow this classloader to be used by the driver and the executor *without* changing `Utils.getSparkClassLoader`.

 at this point am i not reimplementing `spark-submit`? :)

the final approach is certainly worth trying, i'll take a look at it later today.

andrewor14 replied:
We should go with the simplest option that reproduces the issue. In other `SparkSubmitSuite` tests we used (2) but only out of necessity, where we just prepackage a jar and put it in the test resources dir. This makes it a little hard to maintain, e.g. it doesn't work with scala-2.11.

In this case, maybe (3) is the simplest and most maintainable. It's unlikely that we'll ever have to modify `MyException`, but the reproduction code itself should be kept flexible. Could you give it a try?

SparkQA commented:
[Test build #44991 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44991/consoleFull) for   PR 9367 at commit [`53f7c4c`](https://github.com/apache/spark/commit/53f7c4cda8216ce6db726ae88fbf89776da62b27).

SparkQA commented:
[Test build #44991 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44991/console) for   PR 9367 at commit [`53f7c4c`](https://github.com/apache/spark/commit/53f7c4cda8216ce6db726ae88fbf89776da62b27).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44991/
Test FAILed.

srowen commented:
@yhuai @choochootrain how is this one going? it does seem desirable to test this as much as possible. If we're having to introduce very complex mechanisms to do so and they aren't working, is there anything simpler even if less effective we can do to test it?

yhuai commented:
I feel the easiest way is to have something like https://github.com/apache/spark/tree/master/sql/hive/src/test/resources/regression-test-SPARK-8489. So, we will not need to change anything in TestUtils. We just have a jar containing your class and the main object.

choochootrain commented:
@srowen @andrewor14 @yhuai

it seems like the SPARK-8489 approach would be less invasive, but also less maintainable. any preferences? i'd like to stick with one and get this out asap.

brkyvz commented:
My vote is with (3). I feel it requires the least amount of new code that you have to write and is more maintainable.

yhuai commented:
I'd like to go with 3.

Commit:
[SPARK-11195][CORE] Move test to TaskResultGetterSuite

choochootrain commented:
should be less invasive now :)
note that i still have to make some minor changes to `TestUtils` in order for the jar to be in the correct format.

SparkQA commented:
[Test build #45913 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45913/consoleFull) for   PR 9367 at commit [`71f8df9`](https://github.com/apache/spark/commit/71f8df97e6cf3c477101c7fce8c195ae8e0c829f).

SparkQA commented:
[Test build #45913 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45913/console) for   PR 9367 at commit [`71f8df9`](https://github.com/apache/spark/commit/71f8df97e6cf3c477101c7fce8c195ae8e0c829f).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/45913/
Test FAILed.

yhuai commented on core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala:
```diff
@@ -119,5 +124,47 @@ class TaskResultGetterSuite extends SparkFunSuite with BeforeAndAfter with Local
+    // load the exception from the jar
+    val loader = new MutableURLClassLoader(new Array[URL](0), Thread.currentThread.getContextClassLoader)
+    loader.addURL(jarFile.toURI.toURL)
+    Thread.currentThread().setContextClassLoader(loader)
```
Can we set the original loader back?

choochootrain replied:
++

yhuai commented on core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala:
```diff
@@ -119,5 +124,47 @@ class TaskResultGetterSuite extends SparkFunSuite with BeforeAndAfter with Local
+  // SPARK-11195
+  // Make sure we are using the context classloader when deserializing failed TaskResults instead of
+  // the Spark classloader.
+  test("failed task deserialized with the correct classloader") {
```
Can we add comments to outline how this fix is tested?

choochootrain replied:
++

yhuai commented on core/src/main/scala/org/apache/spark/TestUtils.scala:
```diff
@@ -78,15 +79,15 @@ private[spark] object TestUtils {
+   * directory or at the root of the jar.
    */
-  def createJar(files: Seq[File], jarFile: File): URL = {
+  def createJar(files: Seq[File], jarFile: File, directoryPrefix: Option[String] = None): URL = {
```
Do we need to add `directoryPrefix`?

choochootrain replied:
afaik you cannot import a class from a jar unless it has declared a package and the `.class` resides in the corresponding directory in the jar.

without `directoryPrefix`, `createJar` will place all of your classes into the root directory of the jar so you cannot actually import anything from them. it looks like this is the first use of `createJar` in the tests that goes on to load classes from the jar so there was no need for this before.

in the spirit of keeping this change as minimal as possible, i added a `directoryPrefix` option but i am open to other suggestions.

yhuai commented on core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala:
```diff
@@ -17,8 +17,13 @@
 import java.nio.ByteBuffer
 
+import org.apache.spark.TestUtils.JavaSourceFromString
+import org.apache.spark.util.{MutableURLClassLoader, Utils}
```
Let's order imports according to https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Imports.

choochootrain replied:
++

yhuai commented on core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala:
```diff
@@ -119,5 +124,47 @@ class TaskResultGetterSuite extends SparkFunSuite with BeforeAndAfter with Local
+    Thread.currentThread().setContextClassLoader(loader)
+    val excClass: Class[_] = Class.forName("repro.MyException", false, loader)
+
+    sc = new SparkContext("local", "test", conf)
```
I think we need to add a comment to say we need to use `local` in order to make this test work. Otherwise, that jar will not be loaded at the executor side.

choochootrain replied:
++

yhuai commented on core/src/main/scala/org/apache/spark/TestUtils.scala:
```diff
@@ -130,7 +131,7 @@ private[spark] object TestUtils {
     val result = new File(fileName)
     assert(result.exists(), "Compiled file not found: " + result.getAbsolutePath())
-    val out = new File(destDir, fileName)
+    val out = new File(destDir, result.getName)
```
why?

choochootrain replied:
whoops, leftover from the previous changes

Commit:
[SPARK-11195][CORE] Address comments and clean up tests

SparkQA commented:
[Test build #46008 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46008/consoleFull) for   PR 9367 at commit [`fadf2ca`](https://github.com/apache/spark/commit/fadf2ca57e21d5f2600d4ec3e504e20db596c35c).

SparkQA commented:
**[Test build #46008 timed out](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46008/console)**     for PR 9367 at commit [`fadf2ca`](https://github.com/apache/spark/commit/fadf2ca57e21d5f2600d4ec3e504e20db596c35c)     after a configured wait of `175m`.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46008/
Test FAILed.

SparkQA commented:
[Test build #2073 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2073/consoleFull) for   PR 9367 at commit [`fadf2ca`](https://github.com/apache/spark/commit/fadf2ca57e21d5f2600d4ec3e504e20db596c35c).

SparkQA commented:
[Test build #2073 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2073/console) for   PR 9367 at commit [`fadf2ca`](https://github.com/apache/spark/commit/fadf2ca57e21d5f2600d4ec3e504e20db596c35c).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #2079 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2079/consoleFull) for   PR 9367 at commit [`fadf2ca`](https://github.com/apache/spark/commit/fadf2ca57e21d5f2600d4ec3e504e20db596c35c).

SparkQA commented:
[Test build #2080 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2080/consoleFull) for   PR 9367 at commit [`fadf2ca`](https://github.com/apache/spark/commit/fadf2ca57e21d5f2600d4ec3e504e20db596c35c).

SparkQA commented:
[Test build #2079 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2079/console) for   PR 9367 at commit [`fadf2ca`](https://github.com/apache/spark/commit/fadf2ca57e21d5f2600d4ec3e504e20db596c35c).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
[Test build #2080 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2080/console) for   PR 9367 at commit [`fadf2ca`](https://github.com/apache/spark/commit/fadf2ca57e21d5f2600d4ec3e504e20db596c35c).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

yhuai commented:
@choochootrain We should also open a PR for master, right?

yhuai commented on core/src/test/scala/org/apache/spark/scheduler/TaskResultGetterSuite.scala:
```diff
@@ -119,5 +123,64 @@ class TaskResultGetterSuite extends SparkFunSuite with BeforeAndAfter with Local
+    assert(unknownFailure.findFirstMatchIn(exceptionMessage).isEmpty)
+
+    // reset the classloader to the default value
+    Thread.currentThread.setContextClassLoader(originalClassLoader)
```
Can we use the following pattern?
```
// Get the original classloader
try {
// do our test
} finally {
// reset our classloader
}
```
So, we will not mess up the classloader even if the test somehow failed.

choochootrain replied:
++

yhuai commented:
@choochootrain Please open a pr for master branch (this is our typical workflow). So, we can fix it in master and branch-1.6. I'd prefer fixing master and branch 1.6 first and then backport the fix to 1.5.

choochootrain commented:
sounds good, i'll go ahead and squash the commits and submit a pr to master and branch 1.6

yhuai commented:
@choochootrain #9779 has been merged. Can you close this one?

choochootrain commented:
thanks!
